### PR TITLE
keg: installed dependencies of unknown formulae

### LINKED ISSUE #1586: Cannot uninstall dependents unless the dependencies' formulae still exist
CC @alyssais @MikeMcQuaid 

Reproducer:
```
$ brew pull https://github.com/Homebrew/homebrew-core/pull/7288
$ brew install -dvs gtkglext
$ brew install open-scene-graph
$ brew uninstall open-scene-graph
Error: No available formula with the name "pangox-compat"
$ brew uninstall --ignore-dependencies open-scene-graph
Uninstalling /usr/local/Cellar/open-scene-graph/3.5.5... (782 files, 36.9M)
```

As a result, I will now have to revision bump open-scene-graph even though it's not at all broken.

Edit: Note that open-scene-graph does not itself even depend on pangox-compat directly, only via gtkglext, which no longer depends on pangox-compat if 7288 has been pulled. So pangox-compat is not actually anywhere in the dependency tree of open-scene-graph when the error occurs:
```
iMac-TMP:homebrew-core joe$ brew deps open-scene-graph|tr '\n' ' '
atk cairo fontconfig freetype gdk-pixbuf gettext glib gobject-introspection gtk+ gtkglext harfbuzz hicolor-icon-theme icu4c jpeg libffi libpng libtiff openssl pango pcre pixman pkg-config shared-mime-info wget
```

### PR DESCRIPTION
**Pull request by alyssais**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Previously, trying to resolve the dependencies of a keg would raise an exception if the formulae for any of the dependencies could not be found (e.g. if it had been moved to another tap).

This commit updates the dependency finding logic to catch these exceptions, and fall back to comparing names and taps of formulae, which should give the correct behaviour.

Fixes #1586.

### CODE REVIEW

**Comment by alyssais** [comment]
I had to cherry-pick 0a20edf from #1527 to work around #1526.

**Comment by alyssais** [comment]
To test this:

1. Install formula X with a non-core-tap dependency
2. Untap that tap
3. `brew uninstall X`

Previously, this would give you an error. Now, it shouldn't.

File: Library/Homebrew/keg.rb [hunk 1]
```diff
@@ -353,14 +357,21 @@ def to_formula
...
       tab.runtime_dependencies.any? do |dep|
         # Resolve formula rather than directly comparing names
         # in case of conflicts between formulae from different taps.
-        dep_formula = Formulary.factory(dep["full_name"])
-        dep_formula == to_formula && dep["version"] == version.to_s
+        begin
+          dep_formula = Formulary.factory(dep["full_name"])
+          next false unless dep_formula == to_formula
+        rescue FormulaUnavailableError
+          next false unless my_tab["full_name"] = dep["full_name"]
```
Comment (alyssais) [inline Library/Homebrew/keg.rb]: Single equals… oops

**Comment by MikeMcQuaid** [comment]
Thanks @alyssais!

**Pull request merged**